collator/src/collator/do_collate/finalize.rs

```diff
@@ -1628,31 +1636,51 @@ fn split_shard_accounts(
     Ok(shards)
 }
 
-pub fn merge_shard_accounts(
-    shard_accounts: &BTreeMap<u64, ShardAccounts>,
-) -> Result<ShardAccounts> {
-    let shards = shard_accounts.len();
-    let max_depth = shards.trailing_zeros() as u8;
-
-    let mut buffer = shard_accounts
-        .values()
-        .map(|accounts| (max_depth, accounts.clone()))
-        .collect::<VecDeque<_>>();
-
-    while buffer.len() > 1 {
-        let (depth_right, accounts_right) = buffer.pop_back().expect("shouldn't happen");
-
-        match buffer.iter().last() {
-            Some(&(depth_left, _)) if depth_left == depth_right => {
-                let (depth, accounts_left) = buffer.pop_back().expect("shouldn't happen");
-                let merged = accounts_left.merge_with_right_sibling(&accounts_right)?;
-                buffer.push_front((depth - 1, merged));
-            }
-            _ => unreachable!("invalid number of shards"),
+/// Merges multiple shard accounts into a single one using a binary tree approach.
+/// The number of input shards must be equal to `2^SHARD_ACCOUNTS_SPLIT_DEPTH`.
+pub fn merge_shard_accounts(shard_accounts: BTreeMap<u64, ShardAccounts>) -> Result<ShardAccounts> {
+    // Verify input size matches expected number of shards
+    assert_eq!(shard_accounts.len(), 1 << SHARD_ACCOUNTS_SPLIT_DEPTH);
+
+    // Create iterator from BTreeMap
+    let mut iter = shard_accounts
+        .into_values()
+        .map(|accounts| (SHARD_ACCOUNTS_SPLIT_DEPTH, accounts));
+
+    // Initialize array directly from iterator using from_fn
+    let mut buffer: [(u8, ShardAccounts); 1 << SHARD_ACCOUNTS_SPLIT_DEPTH] =
+        std::array::from_fn(|_| iter.next().expect("Length already verified with assert"));
+
+    // Track actual length since we can't truncate arrays
+    let mut buffer_len = 1 << SHARD_ACCOUNTS_SPLIT_DEPTH;
```

Comment:
add a static assert that `sizeof(ShardAccounts) * (1 << SHARD_ACCOUNTS_SPLIT_DEPTH)` is not too big. This mitigates accidental stack overflows if `ShardAccounts` becomes bigger